bonding: Implement active stake checkpointing

victorges opened the pull request:
**What does this pull request do? Explain your changes. (required)**
This is to introduce a checkpointing logic in bonding manager so that we can lookup historical information about any point in time. We need this in order to support an on-chain governance system with support for OpenZeppelin Governor abstractions. That is why it also implements the `IERC5808`  interface to plug into OZ Governor framework.

It is implemented as "external" to `BondingManager` as possible in a separate `BondingVotes` contract, but there were a couple places where `BondingManager` needed changes to start checkpointing its state.

**Specific updates (required)**
- Implement historical active stake snapshotting logic in the `BondingVotes` contract
- Make `BondingManager` call it to checkpoint the state where it needs to
- Tests for the checkpointing logic (unit / integration / fork)

**How did you test each of these updates (required)**
 - Unit and integration tests with `yarn test`
 - Fork test with forge
 - [Devnet on Goerli](https://github.com/livepeer/protocol/pull/620)

**Does this pull request close any open issues?**
Implements PRO-28

**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] README and other documentation updated
- [x] All tests using `yarn test` pass

victorges commented on contracts/bonding/BondingManager.sol:
```diff
@@ -1134,29 +1195,49 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
+     * @return cStake , cFees where cStake is the delegator's cumulative stake including earned rewards and cFees is the
+     * delegator's cumulative fees including earned fees
+     */
+    function delegatorCumulativeStakeAndFees(
```
Should we move this memory/lib version of the func to `EarningsPool(LIP36)?` perhaps?

victorges replied:
Done as it made a lot of sense for testability as well. I didn't want to mock this function just because it happened to be in the `BondingManager`. It makes a lot of conceptual sense as well, since it implements the logic defined in LIP-36, so all the more sense for it to be in `EarningsPoolLIP36`.

victorges commented on test/unit/BondingCheckpoints.js:
```diff
@@ -0,0 +1,1068 @@
+    })
+})
+
+describe("BondingCheckpoints", () => {
```
Realized that these are currently a mix of unit and integration tests. Will be moving and refactoring things around for more well-defined tests

victorges commented on contracts/bonding/BondingCheckpoints.sol:
```diff
@@ -0,0 +1,351 @@
+     */
+    // solhint-disable-next-line func-name-mixedcase
+    function CLOCK_MODE() public pure returns (string memory) {
+        // TODO: Figure out the right value for this
+        return "mode=livepeer_round&from=default";
```
After reading through [this](https://eips.ethereum.org/EIPS/eip-6372#clock_mode):
> If operating using any other mode, then this function SHOULD return a unique identifier for the encoded mode field.

I'll just change this to `mode=livepeer_round` (same thing without "from"). We can discuss in here if there's any debate we might wanna have about it.

victorges commented:
@yondonfu @0xcadams Ok I think NOW this one is ready to go! :)

victorges commented:
Ran the gas report on `confluence` branch (`current`) compared to this (`checkpoints`). Here are the reports, it seems to have increased around 200k gas on the bond/unbond functions due to the checkpoints that have been made. The checkpoint initialization itself also costs around 170k gas, which should still be relatively cheap on Arbitrum, but I haven't found a way to reliably calculate how it would cost considering the calldata cost on L1.

[gas-report-checkpoints.txt](https://github.com/livepeer/protocol/files/12072152/gas-report-checkpoints.txt)
[gas-report-current.txt](https://github.com/livepeer/protocol/files/12072153/gas-report-current.txt)

victorges commented on contracts/bonding/BondingCheckpoints.sol:
```diff
@@ -0,0 +1,393 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+import "@openzeppelin/contracts/utils/Arrays.sol";
```
This is unused in this file, reminder to remove. (used only on `SortedArrays`)

victorges replied:
Not anymore hah

0xcadams commented on contracts/bonding/BondingManager.sol:
```diff
@@ -548,6 +560,40 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
+     * Implemented as a deploy utility to checkpoint the existing state when deploying the BondingCheckpoints contract.
+     * @param _account The account to initialize the bonding checkpoint for
+     */
+    function checkpointBondingState(address _account) external {
```
Doesn't this allow anyone to trigger checkpointing?

victorges replied:
Yeah it does, and that's the idea! The checkpoint process is safe, it will only store the current state of the account in the checkpoints contract (now `BondingVotes`). It will be used during state initialization once we deploy this for the first time, and can be used in the future to correct any potential discrepancies that might show up in the checkpointed state compared to the `BondingManager` state. By having this callable for anyone, we also allow people to make sure other addresses have a consistent checkpoint (or even initialize them on the first round).

0xcadams commented on contracts/bonding/BondingCheckpoints.sol:
```diff
@@ -0,0 +1,393 @@
+
+    /**
+     * @dev Rounds in which we have checkpoints for the total active stake. This and {totalActiveStakeCheckpoints} are
+     * handled in the same wat that {BondingCheckpointsByRound}, with rounds stored and queried on this array and
```
Typo here :)
```suggestion
     * handled in the same way that {BondingCheckpointsByRound}, with rounds stored and queried on this array and
```

0xcadams commented on contracts/bonding/libraries/SortedArrays.sol:
```diff
@@ -0,0 +1,85 @@
+ * @title SortedArrays
+ * @dev Handles maintaining and looking up on sorted uint256 arrays.
+ */
+library SortedArrays {
```
Is there any reusable, audited version of this? Seems like something that should be an openzeppelin contract to reduce surface area

victorges replied:
The reusable version of this is the `Arrays.findUpperBound`, which we do use internally for the more complex logic of actually implementing a binary search. Here we just adapt the "finde upper" to a "find lower" instead which is what we need for the bonding checkpoints. I might be making the interfaces even more similar right next, like making this function return just the index in the same way that OpenZeppelin's interface does it. Will make error checking less redundant in the checkpoint accessing logic.

0xcadams commented on contracts/bonding/BondingCheckpoints.sol:
```diff
@@ -0,0 +1,393 @@
+        virtual
+        returns (uint256 amount, address delegateAddress)
+    {
+        BondingCheckpoint storage bond = getBondingCheckpointAt(_account, _round);
```
I think this can be `memory` vs `storage` to save gas costs

0xcadams replied:
```suggestion
        BondingCheckpoint memory bond = getBondingCheckpointAt(_account, _round);
```

0xcadams replied:
Same across the board for a lot of the view functions

yondonfu replied:
I think using `memory` here would actually increase gas costs because `getBondingCheckpointAt()` currently returns a storage pointer to a struct which I believe defers gas cost consumption for storage reads until the individual fields of the struct are read. If we use `memory` here I believe the values of the storage fields of the struct would need to all be read and written into memory.

In functions where we need to read the values of the struct fields infrequently I think it makes sense to continue using a storage pointer because the gas overhead from reading/copying all struct fields from storage to memory likely outweighs the gas cost of reading the field values direct from storage. In functions where we need to read the values of the struct fields super frequently then it might make sense to read/copy all struct fields from storage to memory because the gas overhead could be small relative to the savings from repeatedly reading from memory instead of storage.

victorges replied:
As a sanity check, I ran the gas reports with [the `memory` version](https://github.com/livepeer/protocol/pull/611/commits/e74f3a4e879dc766b27d4515990f119ba821c93d) for these variables and return values, and the cost of `checkpointBondingState` doesn't change much as expected, while the `getBondingStateAt` function increases gas cost in up to 10k (or 7k on average for delegator/transcoder/nonparticipant). So it is in fact more efficient to keep it as a storage so we don't have to copy the whole struct back to memory.

Attaching the report outputs here FTR.
[gas-memory.txt](https://github.com/livepeer/protocol/files/12361511/gas-memory.txt)
[gas-storage.txt](https://github.com/livepeer/protocol/files/12361512/gas-storage.txt)

0xcadams commented on contracts/bonding/BondingCheckpoints.sol:
```diff
@@ -0,0 +1,393 @@
+    function getBondingCheckpointAt(address _account, uint256 _round)
+        internal
+        view
+        returns (BondingCheckpoint storage)
```
Same here - I *think* this can be memory since this value isn't modified anywhere

0xcadams commented on contracts/bonding/BondingCheckpoints.sol:
```diff
@@ -0,0 +1,393 @@
+    /**
+     * @dev Return IRoundsManager interface
+     */
+    function roundsManager() public view returns (IRoundsManager) {
```
Should this be public vs the internal `boundingManager()` above?

0xcadams reviewed (commented):
Preliminary review of code quality & gas improvements, no deep checks of logic

yondonfu commented on contracts/bonding/BondingCheckpoints.sol:
```diff
@@ -0,0 +1,393 @@
+
+    struct BondingCheckpoint {
+        /**
+         * @dev The amount of bonded tokens to another delegate as per the lastClaimRound.
```
```suggestion
         * @dev The amount of bonded tokens to another delegate as of the lastClaimRound.
```

yondonfu commented on contracts/bonding/BondingManager.sol:
```diff
@@ -524,6 +531,9 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
             if (currPool.cumulativeRewardFactor == 0) {
                 currPool.cumulativeRewardFactor = cumulativeFactorsPool(newDelegate, newDelegate.lastRewardRound)
                     .cumulativeRewardFactor;
+                if (currPool.cumulativeRewardFactor == 0) {
```
Can we document why this line is needed?

victorges replied:
This was meant to account for the case where the delegator bonds to a transcoder that has never called reward. In that case, the `cumulativeRewardsFactor` of the round where the bond was made would stay zeroed and that would affect the pending rewards calculation on the checkpoints contract.

After thinking a bit further about this, also related to the other things we've talked about, I've decided to undo this change to `BondingManager`. Will be good to reduce the amount of changes here and instead have the checkpointing code resilient to such imperfect states that might exist in the bonding manager.

This is more aligned with that idea of not failing on read transactions as much as possible, assuming a default uninitialized state instead. I'm compiling a commit to fix this issue together with the whole mindset shift to return 0 for un-checkpointed accounts or when querying before their first checkpoint. This would make the contract better to use, which I've learned the hard way while trying to interact with the current implementation on the testnet. Will send a detailed explanation of the bigger changes here after I update the PR!

victorges replied:
Reverted this change on https://github.com/livepeer/protocol/pull/614/commits/d5062fe2db7d5037988c2f1a5989886d7ce087d3

yondonfu commented on contracts/bonding/BondingCheckpoints.sol:
```diff
@@ -0,0 +1,393 @@
+         */
+        uint256 lastClaimRound;
+        /**
+         * @dev The last round during which the transcoder called {BondingManager-reward}. This is needed to find a
```
Can we document whether this field is the `lastRewardRound` value for the address that this checkpoint is for vs. the address' delegate?

yondonfu commented on contracts/bonding/BondingCheckpoints.sol:
```diff
@@ -0,0 +1,393 @@
+    /**
+     * @dev See {totalStakeCheckpointRounds} above.
+     */
+    mapping(uint256 => uint256) private totalActiveStakeCheckpoints;
```
A few observations:

- The general abstraction is a wrapper struct like `BondingCheckpointsByRound` that contains a sorted uint256 array of round numbers and a mapping of each round number to the data type that was checkpointed for that round
- The state variables for the above can be kept private

Given the above, rather than maintaining the separate `totalStakeCheckpointRounds` and `totalActiveStakeCheckpoints` state variables and having to maintain the visibility of both (note that `totalStakeCheckpointRounds` is internal by default right now vs. `totalActiveStakeCheckpoints` which is private), what if a single wrapper struct `TotalActiveStakeCheckpointsByRound` that follows the same convention as `BondingCheckpointsByRound` was used? Ex:

```
struct TotalActiveStakeCheckpointsByRound {
    uint256[] rounds;
    mapping (uint256 => uint256) data;
}

TotalActiveStakeCheckpointsByRound private totalActiveStakeCheckpoints;
```

victorges replied:
Liked this very much. Included this fix on https://github.com/livepeer/protocol/commit/d518cd6058e87c7bc10c8ecfc05696e8c30655e8

yondonfu commented on contracts/bonding/BondingCheckpoints.sol:
```diff
@@ -0,0 +1,393 @@
+    /**
+     * @dev Modified to ensure the sender is BondingManager
+     */
+    modifier onlyBondingManager() {
```
Let's move this after the state variable declarations since that has been the convention.

yondonfu commented on contracts/bonding/libraries/SortedArrays.sol:
```diff
@@ -0,0 +1,85 @@
+            uint256 last = array[array.length - 1];
+
+            // values must be pushed in order
+            require(val >= last, "pushSorted: decreasing values");
```
Custom error type instead of require statement here?

yondonfu commented on contracts/bonding/BondingCheckpoints.sol:
```diff
@@ -0,0 +1,393 @@
+     * @dev Helper to return more helpful custom errors in case of bad queries.
+     */
+    function checkedFindLowerBound(uint256[] storage array, uint256 value) internal view returns (uint256) {
+        if (array.length == 0) {
```
Seems redundant if the check is already in `findLowerBound()`.

victorges replied:
Fixed on https://github.com/livepeer/protocol/pull/614/commits/d5062fe2db7d5037988c2f1a5989886d7ce087d3

yondonfu commented on contracts/bonding/BondingCheckpoints.sol:
```diff
@@ -0,0 +1,393 @@
+     */
+    function checkpointTotalActiveStake(uint256 _totalStake, uint256 _round) public virtual onlyBondingManager {
+        if (_round > clock()) {
+            revert FutureCheckpoint(_round, clock());
```
May want to have a separate error type for `checkpointTotalActiveStake()` vs. `checkpointBondingState()` because the requirements for each function are different and it may not be immediately clear just based on the error emitted which requirement was violated. In theory, you could check which function emitted the error, but having separate error types may be cleaner and avoids that extra lookup.

yondonfu commented on contracts/bonding/BondingManager.sol:
```diff
@@ -799,6 +848,8 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
         // Set last round that transcoder called reward
         t.lastRewardRound = currentRound;
 
+        checkpointBondingState(msg.sender, delegators[msg.sender], t);
```
Since `updateTranscoderWithRewards()` calls `increaseTotalStake()` which calls `checkpointBondingState()` for the msg.sender in this function, we have two internal calls to `checkpointBondingState()` for the msg.sender in this function - once before `t.lastRewardRound` is updated to the current round and once after.

If L849 is moved into `updateTranscoderWithRewards()` right before the `increaseTotalStake()` call (L1424) I think it would be possible to remove this extra call to `checkpointBondingState()`.

victorges replied:
Fixed on https://github.com/livepeer/protocol/pull/614/commits/0c41f78401a97ba4d92def55ba21f2c1820ab6b2

On the first implementation I didn't think that this would be a strict requirement, but as I progressed on it and especially after learning about the events interfaces it made a lot more sense to make sure we checkpoint only once per transaction.

The thinking of the original implementation was "whenever the storage is mutated, make sure we checkpoint afterwards" which resulted in the duplicate checkpoints mentioned. I changed this rationale to start from the high-level transaction instead, to make sure that they either checkpoint the state in the end themselves or call some other function (but only once) that checkpoints for them instead.
- Creating a modifier for the checkpointing process made it easier as well so it shows up in the signature of functions that do checkpoint and we can reason about it more quickly.
- There was the need of only 1 function to have an alternate version that doesn't make checkpointing, which I added with the `Uncheckpointed` suffix
- Good thing is I had tests for all those calls already and could verify that all of them were fixed, with only 1 checkpoint per address per tx now

yondonfu commented on contracts/bonding/BondingManager.sol:
```diff
@@ -1441,6 +1484,8 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
         del.bondedAmount = currentBondedAmount;
         del.fees = currentFees;
+
+        checkpointBondingState(_delegator, del, transcoders[_delegator]);
```
This will result in 2 calls to `checkpointBondingState()` in every bonding related function that uses the `autoClaimEarnings()` modifier or the `_autoClaimEarnings()` internal function. Could we add this line to just the end of `claimEarnings()` instead to avoid this?

victorges replied:
Also fixed on https://github.com/livepeer/protocol/pull/614/commits/0c41f78401a97ba4d92def55ba21f2c1820ab6b2 as mentioned in the other comment

yondonfu commented on contracts/bonding/BondingCheckpoints.sol:
```diff
@@ -0,0 +1,393 @@
+        if (array.length == 0) {
+            revert NoRecordedCheckpoints();
+        } else if (array[0] > value) {
+            revert PastLookup(value, array[0]);
```
Should this be folded into SortedArrays since it seems that a bunch of checks are already implemented there?

victorges replied:
Fixed on https://github.com/livepeer/protocol/pull/614/commits/d5062fe2db7d5037988c2f1a5989886d7ce087d3

yondonfu commented on contracts/bonding/BondingManager.sol:
```diff
@@ -548,6 +560,40 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
+    /**
+     * @notice Checkpoints the bonding state for a given account.
+     * @dev This is to allow checkpointing an account that has an inconsistent checkpoint with its current state.
+     * Implemented as a deploy utility to checkpoint the existing state when deploying the BondingCheckpoints contract.
```
Questions:

1. Is there a description of how this is supposed to work?
2. Is the intent to call this function for every address? If so:
    a. Do function calls against the BondingManager need to be prevented until this is complete?
    b. What is the cost estimate?

victorges replied:
I'll answer shortly here for the record, but let's continue the discussion on Discord.

1. I could swear I had included in the spec but apparently not. Need to write this down in some places like LIP and docs.
2. Yeah, to call it for every address that is participating in the protocol, which is around ~4000 from our explorer rn.
   a. They don't! Any action on BondingManager will automatically make a checkpoint as well and has the same end result of calling this function directly.
   b. Around ~200k gas per checkpoint call. The L2 cost seems negligible and something we could pay for, but I don't know much about L1 calldata costs yet.
   
Uncheckpointed accounts used to not be queryable, but after https://github.com/livepeer/protocol/pull/614/commits/d5062fe2db7d5037988c2f1a5989886d7ce087d3 they just return a 0 voting power until checkpointing happens. If the account hasn't been checkpointed, it will have no voting power. So it is in the best interest of the account to have this checkpoint if they intend to vote, and we can even leave for users themselves to call this instead.

So there are some levels of investment we could make which would be fine, just need proper communication with users:
- We can checkpoint all the participant accounts and leave no actions required from users
- We can checkpoint only the active transcoders (they get their whole voting power from that, dont need each delegator to do it) and leave the task for delegators that want to guarantee their voting power to either call this function or make any bonding change (including `claimRewards`, which is the easiest)
- We checkpoint no account ourselves and have an integration in the explorer to call this function if needed
- We checkpoint no account at all and just let users automatically do it when making changes to BondingManager (`reward()`, `claimEarnings()`, etc)

yondonfu commented on contracts/bonding/BondingCheckpoints.sol:
```diff
@@ -0,0 +1,393 @@
+            return bond;
+        }
+
+        if (checkpoints.startRounds.length == 0) {
```
If there are no checkpoints (i.e. startRounds.length == 0) and L255 evaluates to false then it looks like we'll execute L262 which would revert because startRounds.length == 0? This would mean that calling this function for an address without a checkpoint yet (which would be the case immediately post-upgrade for any address that doesn't execute a tx for a bonding action [1]) would cause a revert.

[1] Unless there is a mechanism that guarantees the address has a checkpoint (see my other question about the use of BondingManager.checkpointBondingState()) or that can handle addresses without a checkpoint yet.

victorges replied:
Fixed on https://github.com/livepeer/protocol/pull/614/commits/d5062fe2db7d5037988c2f1a5989886d7ce087d3
Now we just return 0 balance for any uncheckpointed state

victorges commented:
OK @yondonfu @0xcadams I've addressed the feedback on your comments and this is ready for another review round. I have included some other changes as well that seemed necessary after the rounds of testing with the testnet and some discussions.

I recommend reviewing commit by commit, I avoided making a force push here not to lose history, and also isolated the less trivial changes in their own commits. As an overview of each:
- https://github.com/livepeer/protocol/pull/614/commits/0c41f78401a97ba4d92def55ba21f2c1820ab6b2: This addresses the comments about multiple checkpoints being made for the same account in the same transaction. It makes sure we only call it once in the end of all mutations instead on all code paths.
- https://github.com/livepeer/protocol/pull/614/commits/d518cd6058e87c7bc10c8ecfc05696e8c30655e8: This merges the `BondingCheckpointsVotes` contract from https://github.com/livepeer/protocol/pull/615 into the `BondingCheckpoints` contract here (renamed to just `BondingVotes`). This is mainly to allow it to send the `ERC5805` events on changes.
- https://github.com/livepeer/protocol/pull/614/commits/ccc3ddf071a4c7b63fe22ca9531118b858a8b390: This is a change on the way that we assign voting power to accounts, to use their active stake at the **end** of the round instead of **start**. I didn't change any of the core checkpointing logic, just changed the voting power functions to read state from the next round instead, also using the `nextRoundTotalActiveStake` when querying for the next round to the current one.
  - After further testing with the OZ governor, I came to understand that their definition of "voting power at a timepoint" actually refers to the voting power **at the end** of that timepoint. This makes sense if you think about using block numbers as the clock, and is also reflected in the way that the OZ Governor works for some things like vote snapshotting: voting starts only **after** the "snapshot round", not **on** it.
  - A practical effect of this is that, in the previous implementation, if we had a `votingDelay` of 1 round the voting would be delayed in 1-2 rounds, but users would only really have 0-1 rounds to make any staking changes. Any staking changes made on the round before the voting actually starts would not be effective, since we would take the active stake from the start of that round for the voting power. Now we take it from the end (start of next round), and things work a bit more as expected.
- https://github.com/livepeer/protocol/pull/614/commits/d5062fe2db7d5037988c2f1a5989886d7ce087d3: This changes the behavior of most of the "read" logic in the `BondingVotes` contract to avoid reverting queries that could simply return a 0 value. There is only 1 specific case now where we will revert instead of returning zero, which is when there is no earnings pool on a transcoder's last reward round, which is an illegal state from the way that `BondingManager` works. This also fixed a lot of the redundant `reverts` with the `SortedArrays` implementation.
- https://github.com/livepeer/protocol/pull/614/commits/292f9542610d6f270966570135915a12ce6fd531: This one addresses the other review comments which were simpler so I just combined them in this commit.

victorges commented on contracts/test/mocks/BondingVotesERC5805Harness.sol:
```diff
@@ -0,0 +1,34 @@
+ * @dev This is a tets utility for unit tests on the ERC5805 functions of the BondingVotes contract. It overrides the
+ * functions that should be used to derive the values returned by the ERC5805 functions and checks against those.
+ */
+contract BondingVotesERC5805Harness is BondingVotes {
```
So the idea here is that the `BondingVotes` implementation has 2 "layers". The first layer deals with checkpointing directly, while the second layer is the logic that was in the old `BondingCheckpointsVotes`. So the tests for that second layer are simpler, just checking that it proxies to the checkpointing logic.

This "harness" works by mocking only that first layer of the implementation, then checking if the second layer, the ERC5805 functions, are depending only on it.

Let me know if you think this testing strategy is too confusing and we need more guarantees here. One idea I tried was creating a separate abstract `BondingVotesERC5805` contract, which got inherited by the main implementation. We could have some more isolated/guaranteed tests and clearer implementation, but it felt like a little overengineering and I reverted that. Let me know if it'd be preferrable, or if you have any other ideas. Should be easy to adapt here.

yondonfu replied:
Not sure I understand how this is being used from a quick glance. If you link to the relevant areas that this contract is used and the test cases written for it can prob comment on that.

victorges replied:
This contract is used in [ERC5805 unit tests here](https://github.com/livepeer/protocol/blob/1692ec11e1cf29d3a26f29092b9e9fe12890b51d/test/unit/BondingVotes.js#L1161).

What feels wrong there is that even though it is a unit test of the `BondingVotes` contract, those tests actually replace the unit under test with another one 😅 
It makes sense for the current version of the code, but it could get confusing when someone else looks at this or maybe makes changes to the "ERC5805-layer".

Separating it to an actual separate "module" on the implementation would allow for a more proper unit test, which could create the "harness" only surrounding that module in a separate unit test [1] (like [on treasury counting module](https://github.com/livepeer/protocol/pull/615/files#diff-ad14a7fcaebacc5393295641e68e2eee7817d93aa51ecbeaeffc5e2f14e2d96aR16)), instead of replacing the whole `BondingVotes` under test with a contract inheriting from it that overrides a couple functions. This is the path that also feels like overengineering though, so I left it like this for now.

Let me know if you have a preference for the separate module and I can change it if so.

[1] e.g.:
```
abstract contract BondingVotesERC5805 is IVotes {
   // implement ERC5805 functions 
   // requires implementation of `clock()`, `getBondingStateAt` and `getTotalActiveStakeAt`
}

contract BondingVotes is IBondingVotes, BondingVotesERC5805 {
  // implements the core checkpointing logic incl the abstract functions that BondingVotesERC5805 wants
}
```
like this we could test `BondingVotesERC5805` more isolatedly (tho no actual difference in the tests themselves, only on the setup)

yondonfu replied:
Ah I see. 

Hmm it seems that the bulk of the core logic in BondingVotes right now *is* the checkpointing logic and then all the ERC-5805 functions are either fairly simple or rely on calling functions that implement the checkpointing logic. While the module approach might make testing a bit easier, it does feel like it would end up breaking up into a module (i.e. BondingVotesERC5805) that doesn't do much that just relies on abstract functions that need to be implemented in the child contract. I feel ok about the current testing approach..., but I would just document the rationale in the tests themselves that use the harness contract.

yondonfu commented on contracts/bonding/BondingManager.sol:
```diff
@@ -123,6 +124,11 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
         _;
     }
 
+    modifier autoCheckpoint(address account) {
```
```suggestion
    modifier autoCheckpoint(address _account) {
```

to follow convention elsewhere.

yondonfu commented on contracts/bonding/BondingManager.sol:
```diff
@@ -1378,11 +1427,14 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
     /**
      * @dev Update a delegator with token pools shares from its lastClaimRound through a given round
+     *
+     * Notice that this function udpates the delegator storage but does not checkpoint its state. Since it is internal
```
```suggestion
     * Notice that this function updates the delegator storage but does not checkpoint its state. Since it is internal
```

yondonfu commented on contracts/bonding/BondingManager.sol:
```diff
@@ -546,6 +559,43 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
+    /**
+     * @notice Checkpoints the bonding state for a given account.
+     * @dev This is to allow checkpointing an account that has an inconsistent checkpoint with its current state.
+     * Implemented as a deploy utility to checkpoint the existing state when deploying the BondingVotes contract.
```
Based on latest discussion on how checkpointing will work this line seems like it can be removed.

yondonfu commented on contracts/bonding/BondingManager.sol:
```diff
@@ -546,6 +559,43 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
+     * @notice Checkpoints a delegator state after changes, to be used for historical voting power calculations in
+     * on-chain governor logic.
+     */
+    function checkpointBondingState(
```
The convention has been for internal functions to be grouped together later in the file after public functions.

victorges replied:
Oh cool, I hadn't noticed this was the pattern. Will update it here and might need in some other places like the `_setTreasuryRewardCut` internal function on https://github.com/livepeer/protocol/pull/616

While at it, should internal functions have a `_` prefix as well? What is the current pattern? I couldn't tell from the existing functions. In the case of `_setTreasuryReward` cut it needs it not to name-clash with the external one tho

yondonfu replied:
IIRC the convention has been to have `_` prefix if the internal function signature would clash with the public function signature. I recall internal functions getting names that describe the logic they contain and in the scenario that the internal function contains almost exactly the same logic as a corresponding public function then the internal function gets the public function name with the `_` prefix.

victorges replied:
OK, I think I'll rename the internal function to have the `_` prefix then. Even though the signature doesn't clash cause the arguments are different, I think it's worth the differentiation for clarity. Will also make the `public` one `external` and call the internal one directly from the `modifier`

yondonfu commented on contracts/bonding/BondingManager.sol:
```diff
@@ -546,6 +559,43 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
         emit Bond(_to, currentDelegate, _owner, _amount, del.bondedAmount);
+
+        // the `autoCheckpoint` modifier has been replaced with its internal function as a `Stack too deep` error work-around
+        checkpointBondingState(_owner, del, transcoders[_to]);
```
Shouldn't the last arg be diff?

```suggestion
        checkpointBondingState(_owner, del, transcoders[_owner]);
```

victorges replied:
Wow yeah. I'll make sure to add some test checks to ensure this. The only reason it didn't affect them is because we only use the `lastRewardRound` from the `Transcoder` struct right now, which doesn't affect the bonding state calculation for delegators (cause we check the `delegateAddress` before that).

victorges replied:
Addressed on https://github.com/livepeer/protocol/pull/614/commits/fecf5a2e0e0cf76472ef1c23d543bc6f9512fb9f together with the test updates

yondonfu commented on contracts/bonding/BondingVotes.sol:
```diff
@@ -0,0 +1,534 @@
+     * @dev Return BondingManager interface
+     */
+    function bondingManager() internal view returns (BondingManager) {
+        return BondingManager(controller.getContract(keccak256("BondingManager")));
```
Can this be an IBondingManager interface?

victorges replied:
It currently cannot as it doesn't include the 2 functions that we call from it: `nextRoundTotalActiveStake` and `getTranscoderEarningsPool`

What's the process with these interfaces? Should I just add the methods that I need to the interface? They generally don't seem to be extensive of the contracts public functions.

yondonfu replied:
> What's the process with these interfaces?

In the past, we've just added the external/public functions that need to be called by other protocol contracts.

I believe it would be beneficial to use the IBondingManager interface here if possible as it would avoid importing the entirety of BondingManager.sol which I *think* affects the deployment bytecode size.

victorges replied:
OK, will do. 

As for the interface I created here, `IBondingVotes`, I've actually included all of the public functions that it exports. Let me know if it'd be preferrable to trim that interface down to only the methods used externally (likely only the checkpoint ones used by `BondingManager`, since the treasury uses only the `IVotes` interface directly.

Should events/errors also be only in the implementation or kept in the interface? Or keep in the interface only the ones emitted from functions in the interface?

yondonfu replied:
> As for the interface I created here, IBondingVotes, I've actually included all of the public functions that it exports

I think that is fine and probably the overall direction to go in moving forward for the reasoning I mention below (though don't think its necessary to update existing interfaces that are already being used immediately to follow that convention and can get to that separately i.e. IBondingManager).

> Should events/errors also be only in the implementation or kept in the interface? Or keep in the interface only the ones emitted from functions in the interface?

I think having events/errors in the interface is a good idea as it makes it easier to just look at the interface contract and know what a consumer of the contract has access to/needs to be aware of in the public API.

yondonfu reviewed (commented):
Still need to do the follow up for BondingVotes.sol.

yondonfu commented on contracts/bonding/BondingVotes.sol:
```diff
@@ -0,0 +1,534 @@
+     * @notice Returns the name of the virtual token implemented by this.
+     */
+    function name() external pure returns (string memory) {
+        return "Livepeer Stake";
```
Recommend something more descriptive of what the virtual token is used for like "Voting LPT".

victorges replied:
That makes sense. How about `Livepeer Voting Power`?

yondonfu replied:
SGTM

yondonfu commented on contracts/bonding/BondingVotes.sol:
```diff
@@ -0,0 +1,534 @@
+     * @notice Returns the delegate that _account had chosen in a specific round in the past. See `delegates()` above
+     * for more details.
+     * @dev This is an addition to the IERC5805 interface to support our custom vote counting logic that allows
+     * delegators to override their transcoders votes. See {GaovernorCountingOverridable-_handleVoteOverrides}.
```
```suggestion
     * delegators to override their transcoders votes. See {GovernorCountingOverridable-_handleVoteOverrides}.
```

yondonfu commented on contracts/bonding/BondingVotes.sol:
```diff
@@ -0,0 +1,534 @@
+
+    /**
+     * @notice Returns whether an account already has any checkpoint.
+     * @dev This is meant to be called by a checkpoint initialization script once we deploy the checkpointing logic for
```
Seems that this line can be removed given latest thinking on checkpoint initialization by voters prior to voting.

victorges commented on contracts/bonding/IBondingVotes.sol:
```diff
@@ -0,0 +1,48 @@
+/**
+ * @title Interface for BondingVotes
+ */
+interface IBondingVotes is IERC6372Upgradeable, IVotes {
```
Self-reminder: `IERC6372Upgradeable` is redundant here since it's included in `IVotes`. Remove

yondonfu commented on contracts/bonding/BondingVotes.sol:
```diff
@@ -0,0 +1,534 @@
+            emit DelegateVotesChanged(_account, previous.delegatedAmount, 0);
+        }
+
+        // Always send delegator events since transcoders are delegators themselves. The way our rewards work, the
```
This event name feels misleading as the event doesn't actually get emitted every time a delegator's voting power changes.

victorges replied:
Makes sense. I can move it to the `IBondingVotes` interface (as opposed to the "generic" `IVotes`) and rename it to `DelegatorBondedAmountChanged`, since it is an internal thing from our abstractions anyway

yondonfu commented on contracts/bonding/BondingVotes.sol:
```diff
@@ -0,0 +1,534 @@
+     * @param _round The round for which the total active stake is valid. This is normally the current round.
+     */
+    function checkpointTotalActiveStake(uint256 _totalStake, uint256 _round) public virtual onlyBondingManager {
+        if (_round > clock()) {
```
Should this be `_round != clock() + 1` to mirror the conditional in `checkpointBondingState()`?

victorges replied:
I've changed this to `_round != clock()` and renamed the error to `InvalidTotalStakeCheckpointRound`.

The checkpointing rounds are actually different between individual accounts and total active stake tho. While individual accounts are always checkpointed on the next round, from the way that active stake update works on BondingManager, the total active stake is checkpointed for the current round instead, when the round is initialized and `nextRoundTotalActiveStake` becomes the current.

We can see it as having an optimization for the total active stake: since we have that single point where we update 1 variable for the total stake, we can leave the checkpointing for that time only. That is cheaper than checkpointing the `nextRoundTotalActiveStake` for the next round every time it changes, which would be the analogous flow to the individual accounts checkpoints.

yondonfu commented on contracts/bonding/BondingVotes.sol:
```diff
@@ -0,0 +1,534 @@
+     * @notice Returns the symbol of the token underlying the voting power.
+     */
+    function symbol() external view returns (string memory) {
+        return livepeerToken().symbol();
```
Recommend using a different symbol to distinguish from LPT i.e. "vLPT".

yondonfu commented on contracts/bonding/BondingVotes.sol:
```diff
@@ -0,0 +1,534 @@
+     * @notice Returns the decimals places of the token underlying the voting.
+     */
+    function decimals() external view returns (uint8) {
+        return livepeerToken().decimals();
```
Should be able to avoid referencing the LPT contract here by just returning 18 - doesn't seem necessary to call the LPT contract here.

yondonfu commented on contracts/bonding/BondingVotes.sol:
```diff
@@ -0,0 +1,534 @@
+            uint256 nextInitedRound = initializedRounds[upper];
+            return totalStakeCheckpoints.data[nextInitedRound];
+        } else {
+            // Here the _round is after any initialized round, so grab its stake from nextRoundTotalActiveStake()
```
Shouldn't ever get here right due to L325?

victorges replied:
We actually might in the following cases:
 - The current round has not been initialized yet, so instead of reverting or returning an inaccurate value until the round-initializer gets to it, we can just return `nextRoundTotalActiveStake`
 - This is being called for the next round (e.g. when querying for the current `totalSupply()`), in which case the checkpoint will never be present since we only checkpoint it on the round initialization
 - A couple rounds have passed without having been initialized. All those rounds should use the `nextRoundTotalActiveStake` value as well to be internally consistent with the values we return for `getBondingCheckpointAt`. 
 
 This is related to that different behavior between individual & total checkpoints, explained in the other comment https://github.com/livepeer/protocol/pull/614#discussion_r1303341851
If we checkpointed the total active stake the "expensive way", every time `nextRound...` is changed, we could have the same logic for querying here: grab the lower bound checkpoint.
Since we checkpoint only on the next round instead, it means that the "uncheckpointed total stake value" stays in the `BondingManager` contract until another round is initialized. Which is why here we:
- Use "upper bound" instead of "lower bound" to find the stake of a previous round
- Use `nextRoundTotalActiveStake` if there are no rounds initialized after the searched round

victorges commented on contracts/bonding/BondingVotes.sol:
```diff
@@ -0,0 +1,534 @@
+     * terms), we need to fetch the bonding state at the next round instead. That because the bonding state reflects the
+     * active stake in the current round, which is the snapshotted stake from the end of the previous round.
+     */
+    function getPastVotes(address _account, uint256 _round) public view returns (uint256) {
```
FTR of an offline discussion: The ERC5805 specifies that the `getPast...` functions should only be callable with a past timepoint value, so that its results can always be immutable (compared to `getVotes` that returns votes in current round [1]).

So we decided to make that change here and add an additional revert in case this is called for the current round. State from the current round 

[1] This behavior feels even more aligned with the change to return the stake in the end of the round instead of the start. The `getVotes` **is** actually mutable now, which it wasn't when we were returning the stake at the beginning of the round.

victorges commented on src/test/BondingVotesStateInitialization.sol:
```diff
@@ -0,0 +1,278 @@
+
+        address[2] memory testAddresses = [DELEGATOR, TRANSCODER];
+        for (uint256 i = 0; i < testAddresses.length; i++) {
+            CHEATS.expectRevert(IBondingVotes.NoRecordedCheckpoints.selector);
```
These tests are outdated, update

victorges commented on contracts/bonding/BondingVotes.sol:
```diff
@@ -0,0 +1,534 @@
+
+        bool isTranscoder = newDelegate == _account;
+        bool wasTranscoder = previousDelegate == _account;
+        if (isTranscoder) {
+            emit DelegateVotesChanged(_account, previous.delegatedAmount, current.delegatedAmount);
+        } else if (wasTranscoder) {
+            // if the account stopped being a transcoder, we want to emit an event zeroing its "delegate votes"
+            emit DelegateVotesChanged(_account, previous.delegatedAmount, 0);
+        }
```
Improving this part a bit to only send these events when the `delegatedAmount` actually changed.

yondonfu commented on contracts/bonding/BondingVotes.sol:
```diff
@@ -0,0 +1,558 @@
+
+    /**
+     * @notice Returns the current total supply of votes available.
+     * @dev This value is the sum of all *active* stake, which is not necessarily the sum of all voting power.
```
After thinking on this a bit, this feels weird.

I would expect that totalSupply() == sum(voting power). While not having this invariant doesn't seem like it would break anything, not having this invariant in place feels like smell...if I query totalSupply() and then index all votes for a proposal I would expect to be able to calculate the % participation in a vote as sum(voting power indexed) / totalSupply(), but that would not be wrong because the denominator is incorrect.

I know the current offchain poll tallying logic has the same problem, but if the lift to correct this problem is small (keeping track of an additional state variable in checkpoints to determine if an address was either an active transcoder or delegated to an active transcoder as of the timepoint?) to I would lean towards fixing it to start off on solid footing with proper invariants in place for onchain voting.

yondonfu replied:
Discussed offline addressing this separately from this PR

victorges replied:
FTR we discussed offline and concluded the current behavior is reasonable given the current audit contest timelines. Might followup with another PR changing this behavior if there's time, but focusing instead on documenting the current behavior and contest for now.

edit: always surprised how github doesnt autorefresh the page

yondonfu reviewed (approved):
LGTM
